gateway/src/services/config-manager-v2.ts

```diff
@@ -3,29 +3,60 @@ import { ValidateFunction } from 'ajv';
 import fs from 'fs';
 import path from 'path';
 import yaml from 'js-yaml';
+import * as migrations from '../../conf/migration/migrations';
 
 type Configuration = { [key: string]: any };
 type ConfigurationDefaults = { [namespaceId: string]: Configuration };
+export type Migration = (
+  configRootFullPath: string,
+  configRootTemplateFullPath: string
+) => void;
+type MigrationFunctions = {
+  [key: string]: Migration;
+};
 interface _ConfigurationNamespaceDefinition {
   configurationPath: string;
   schemaPath: string;
 }
 type ConfigurationNamespaceDefinition = _ConfigurationNamespaceDefinition & {
   [key: string]: string;
 };
-type ConfigurationRoot = {
+type ConfigurationNamespaceDefinitions = {
   [namespaceId: string]: ConfigurationNamespaceDefinition;
 };
+interface ConfigurationRoot {
+  version: number;
+  configurations: ConfigurationNamespaceDefinitions;
+}
 const NamespaceTag: string = '$namespace ';
 const ConfigRootSchemaPath: string = path.join(
   __dirname,
   'schema/configuration-root-schema.json'
 );
+const ConfigTemplatesDir: string = path.join(__dirname, '../templates/');
 interface UnpackedConfigNamespace {
   namespace: ConfigurationNamespace;
   configPath: string;
 }
 
+export function deepCopy(srcObject: any, dstObject: any): any {
+  for (const key in srcObject) {
+    if (typeof srcObject[key] !== 'object') {
+      if (typeof dstObject[key] !== 'object') dstObject[key] = srcObject[key];
+    } else {
+      if (!dstObject[key]) {
+        if (srcObject[key] instanceof Array) dstObject[key] = [];
```

Comment:
WTF? Have you actually tested this function yourself?? This doesn't work for arrays since `for (const key in srcObject)` with an array would iterate the values instead of keys.

Reply:
Change the structure of this function to this, to make it easier to read.

```
if (srcObject instanceof Array) {
  ...
} else if (srcObject instanceof Object) {
  ...
} else {
  ...
}
```

Reply:
And make sure to include a unit test with some nested objects and arrays. e.g.

```
{
  key1: "string value",
  key2: {
    ...
  },
  key3: [
    { ... },
    { ... },
    ...
  ],
}
```

Reply:
A unit test is already included in this pr to test similar case - https://github.com/CoinAlpha/hummingbot/blob/6efa89f5662419b87c594078c88a48d16b1a49fa/gateway/test/services/config-manager-v2.test.ts#L163
Can you take a look?

Reply:
Also, Typescript seems to have issue with this `for (const [key, value] in Object.entries(srcObject)) {`. I'll try to figure a solution to that.

Reply:
You have a unit test for having objects and arrays, but not nested objects and arrays. What I mean is you don't have cases like "key3" above - where you have objects inside arrays. The converse (arrays inside something like key2) should also be tested.

Reply:
Done!

Reply:
One thing to note is that the order of entries in arrays are sensitive during deepcopy. Inserting new element at the beginning or in between an array list will cause entries after that insertion to be overwritten.

Reply:
I also think there's really no need to cater for such cases as that will make the  logic very complex.